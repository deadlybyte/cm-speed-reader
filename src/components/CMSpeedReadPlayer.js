// @flow
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SpeedyReader from 'react-speedy-reader';
import fullTimeWhistleAudio from '../media/audio/full-time-whistle.mp3';

class CMSpeedReadPlayer extends Component {
  static defaultProps = {
    isMuted: true
  };

  static propTypes = {
    isMuted: PropTypes.bool,
    speed: PropTypes.number.isRequired
  };

  constructor(props) {
    super(props);
    this.playFinalWhistle = this.playFinalWhistle.bind(this);
  }

  async playFinalWhistle() {
    const { isMuted } = this.props;

    if (!isMuted) {
      const audio = new Audio(fullTimeWhistleAudio);
      await audio.play();
    }
  }

  render() {
    const { playFinalWhistle } = this;
    const { speed } = this.props;

    return (
      <div className="CM-block CM-commentary-block">
        <p className="App-intro">
          <SpeedyReader
            autoPlay
            inputText="To get started, edit src/App.js and save to reload."
            speed={speed * 100}
            onFinish={playFinalWhistle}
          />
        </p>
      </div>
    );
  }
}

export default CMSpeedReadPlayer;
