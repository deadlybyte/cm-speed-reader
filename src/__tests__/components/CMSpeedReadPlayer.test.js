import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import CMSpeedReadPlayer from '../../components/CMSpeedReadPlayer';

describe('CMSpeedReadPlayer', () => {
  describe('render', () => {
    it('should render without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<CMSpeedReadPlayer speed={1} />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    it('should render as expected', () => {
      const componentUnderTest = renderer.create(<CMSpeedReadPlayer speed={1} />)
          .toJSON();
        expect(componentUnderTest).toMatchSnapshot();
    });
  });

  describe('playFinalWhistle', () => {
    let playMock;

    beforeEach(() => {
      playMock = jest.spyOn(Audio.prototype, 'play');
    });

    afterEach(() => {
      playMock.mockClear();
    });

    it('should play audio when passage has been read and audio not muted', () => {
      const componentUnderTest = shallow(<CMSpeedReadPlayer isMuted={false} speed={1} />);
      componentUnderTest.instance().playFinalWhistle();

      expect(playMock).toHaveBeenCalled();
    });

    it('should not play audio when passage has been read by default', () => {
      const componentUnderTest = shallow(<CMSpeedReadPlayer speed={1} />);
      componentUnderTest.instance().playFinalWhistle();

      expect(playMock).not.toHaveBeenCalled();
    });

    it('should not play audio when passage has been read and audio is muted', () => {
      const componentUnderTest = shallow(<CMSpeedReadPlayer isMuted speed={1} />);
      componentUnderTest.instance().playFinalWhistle();

      expect(playMock).not.toHaveBeenCalled();
    });
  });
});
