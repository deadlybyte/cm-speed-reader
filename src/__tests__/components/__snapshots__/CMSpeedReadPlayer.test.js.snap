// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CMSpeedReadPlayer render should render as expected 1`] = `
<div
  className="CM-block CM-commentary-block"
>
  <p
    className="App-intro"
  >
    <span>
      Â 
    </span>
  </p>
</div>
`;
